src/hooks/useTaskRunsV2.ts

```diff
@@ -0,0 +1,170 @@
+import * as React from 'react';
+import { uniqBy } from 'lodash-es';
+import { useIsOnFeatureFlag } from '~/feature-flags/hooks';
+import { WatchK8sResource } from '~/types/k8s';
+import { useK8sWatchResource } from '../k8s';
+import { useKubearchiveListResourceQuery } from '../kubearchive/hooks';
+import { TaskRunGroupVersionKind, TaskRunModel } from '../models';
+import { TaskRunKind } from '../types';
+import { createKubearchiveWatchResource } from '../utils/task-run-filter-transforms';
+import { GetNextPage, NextPageProps, useTRTaskRuns } from './useTektonResults';
+// eslint-disable-next-line import/order
+import { TektonResultsOptions } from '~/utils/tekton-results';
+
+/**
+ * Hook for fetching TaskRuns with feature flag controlled data source switching.
+ *
+ * When 'taskruns-kubearchive' feature flag is enabled, combines live cluster data
+ * via useK8sWatchResource with KubeArchive historical data for complete coverage.
+ * Otherwise, uses the similar implementation for tekton results.
+ *
+ * Returns interface matches existing useTaskRuns exactly with infinite loading support.
+ */
+export const useTaskRunsV2 = (
+  namespace: string,
+  options?: Partial<Pick<WatchK8sResource, 'watch' | 'limit' | 'selector' | 'fieldSelector'>>,
+): [TaskRunKind[], boolean, unknown, GetNextPage, NextPageProps] => {
+  const enableKubearchive = useIsOnFeatureFlag('taskruns-kubearchive');
+
+  // Live cluster data - always fetched for consistency
+  const {
+    data: clusterResources,
+    isLoading: clusterLoading,
+    error: clusterError,
+  } = useK8sWatchResource<TaskRunKind[]>(
+    namespace
+      ? {
+          groupVersionKind: TaskRunGroupVersionKind,
+          namespace,
+          isList: true,
+          selector: options?.selector,
+          fieldSelector: options?.fieldSelector,
+          watch: options?.watch !== false,
+        }
+      : null,
+    TaskRunModel,
+    { retry: false },
+  );
+
+  // Process cluster data with sorting (no limit applied yet)
+  const processedClusterData = React.useMemo(() => {
+    if (clusterLoading || clusterError || !clusterResources) return [];
+
+    const sorted = [...clusterResources].sort((a, b) =>
+      (b.metadata?.creationTimestamp || '').localeCompare(a.metadata?.creationTimestamp || ''),
+    );
+
+    return sorted;
+  }, [clusterResources, clusterLoading, clusterError]);
+
+  // should query when there is no limit or no enough data in cluster or cluster meets error
+  const needsMoreData =
+    !clusterLoading &&
+    (!options?.limit || // no limit case
+      (clusterResources && options.limit > clusterResources.length) || // not enough data
+      !!clusterError); // error after load
+
+  const shouldQueryTekton = !enableKubearchive && needsMoreData;
```

Comment:
nit: you can also check for `&& namespace` here (and for `shouldQueryKubearchive` as well)